test/parallel-benchmark/mzcompose.py

```diff
@@ -0,0 +1,1186 @@
+
+
+class Action:
+    def run(
+        self,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ):
+        start_time = datetime.datetime.now()
+        self._run(conns)
+        duration = datetime.datetime.now() - start_time
+        measurements[str(self)].append(
+            Measurement(duration.total_seconds(), start_time)
+        )
+
+    def _run(self, conns: queue.Queue):
+        raise NotImplementedError
+
+
+class TdAction(Action):
+    def __init__(self, td: str, c: Composition):
+        self.td = dedent(td)
+        self.c = c
+
+    def _run(self, conns: queue.Queue):
+        self.c.testdrive(self.td, quiet=True)
+
+    def __str__(self) -> str:
+        return "testdrive"
+
+
+class StandaloneQuery(Action):
+    def __init__(
+        self,
+        query: str,
+        conn_info: PgConnInfo,
+        strict_serializable: bool = True,
+    ):
+        self.query = query
+        self.conn_info = conn_info
+        self.strict_serializable = strict_serializable
+
+    def _run(self, conns: queue.Queue):
+        conn = self.conn_info.connect()
+        with conn.cursor() as cur:
+            if not self.strict_serializable:
+                cur.execute("SET TRANSACTION_ISOLATION TO 'SERIALIZABLE'")
+            cur.execute(self.query)
+        conn.close()
+
+    def __str__(self) -> str:
+        return f"{self.query} (standalone)"
+
+
+class ReuseConnQuery(Action):
+    def __init__(
+        self,
+        query: str,
+        conn_info: PgConnInfo,
+        strict_serializable: bool = True,
+    ):
+        self.query = query
+        self.conn = conn_info.connect()
+        self.conn.autocommit = True
+        self.cur = self.conn.cursor()
+        if not strict_serializable:
+            self.cur.execute("SET TRANSACTION_ISOLATION TO 'SERIALIZABLE'")
+
+    def _run(self, conns: queue.Queue):
+        self.cur.execute(self.query)
+
+    def __str__(self) -> str:
+        return f"{self.query} (reuse connection)"
+
+
+class PooledQuery(Action):
+    def __init__(self, query: str, conn_info: PgConnInfo):
+        self.query = query
+        self.conn_info = conn_info
+
+    def _run(self, conns: queue.Queue):
+        conn = conns.get()
+        with conn.cursor() as cur:
+            cur.execute(self.query)
+        conns.task_done()
+        conns.put(conn)
+
+    def __str__(self) -> str:
+        return f"{self.query} (pooled)"
+
+
+class Distribution:
+    def generate(self, duration: int) -> Iterator[None]:
+        raise NotImplementedError
+
+
+class Periodic(Distribution):
+    """Run the action in each thread in one second, spread apart by the 1/per_second"""
+
+    def __init__(self, per_second: int):
+        self.per_second = per_second
+
+    def generate(self, duration: int) -> Iterator[None]:
+        sleep_time = 1 / self.per_second
+        for i in range(duration * self.per_second):
+            yield
+            time.sleep(sleep_time)
```

Comment:
This doesn't produce a true periodic distribution as it can be affected by scheduling. Instead, it should determine at which point it needs to produce a record, and sleep the right amount of time. This ensures that even under load, we'll produce requests that are in the limit evenly-spaced.